Database/DB Connection Pool.md

```diff
@@ -0,0 +1,127 @@
+
+Nginx 에러 로그, PM2 에러 로그, winston을 이용한 에러로그 확인을 모두 했으나 뾰족하게 원인을 찾기 힘들었다. 이곳 저곳 물어보다가 결국 `DB Connection Pool` 을 Release 하지 않아서 발생한 문제라는 것을 알게 됐다. 
+
+구글링을 해보니 `DB Connection Pool` 문제로 인해 서버 오류가 터졌을 때 그 원인이 `DB Connection Pool` 이라는 것을 단번에 찾기 어렵다는 글이 많았다. ~~(휴 발견해서 다행 ㅋ)~~
+
+그래서 정리해보려 한다. `DB Connection Pool` 은 무엇이며 `Pool`을 `Release` 해준다는 것은 어떤 의미인가? 만약 `Pool`은 여러 개 생성만 하고 `Release`를 하지 않으면 어떤 문제가 발생하는가? 알아보자.
+
+# DB Connection Pool 이란?
+
+### Single Connection
+
+`Single Connection`을 사용할 경우 해당 `Connection`에 여러 `Statement`를 사용하게 된다. 여러 `Statement` 중 한 `Statement`에서 예외가 발생하면 `Rollback`을 수행해야 하는데 `Single Connection`의 경우는 해당 `Connection`을 통해 생성된 다른 모든 `Statement`의 작업도 `Rollback`되게 된다. (원치 않는 작업이 발생한다)
+
+> [npm mysql2](https://www.npmjs.com/package/mysql2#using-connection-pools)
+
+```jsx
+// get the client
+const mysql = require('mysql2');
+ 
+// create the connection to database
+const connection = mysql.createConnection({
+  host: 'localhost',
+  user: 'root',
+  database: 'test'
+});
+ 
+// simple query
+connection.query(
+  'SELECT * FROM `table` WHERE `name` = "Page" AND `age` > 45',
+  function(err, results, fields) {
+    console.log(results); // results contains rows returned by server
+    console.log(fields); // fields contains extra meta data about results, if available
+  }
+);
+ 
+// with placeholder
+connection.query(
+  'SELECT * FROM `table` WHERE `name` = ? AND `age` > ?',
+  ['Page', 45],
+  function(err, results) {
+    console.log(results);
+  }
+);
+```
+
+참고> 위 예시는 날 쿼리를 그대로 사용했기 때문에 [SQL Injection](https://ko.wikipedia.org/wiki/SQL_%EC%82%BD%EC%9E%85)의 위험이 있을 수 있다. 그래서 [Prepared Statement](https://www.npmjs.com/package/mysql2#using-prepared-statements)라는 것을 대신 사용한다. 이유는 [이 사이트](https://stackoverflow.com/questions/8263371/how-can-prepared-statements-protect-from-sql-injection-attacks)를 참조.
+
+### 단순 Multiple Connection
+
+![2](https://user-images.githubusercontent.com/43839938/97099039-537dfa00-16c7-11eb-8b24-6f9b9bbe3fa7.png)
+
+
+위의 문제를 해결하려면 클라이언트의 각 요청에 대해 개별 `Connection`을 사용해야 한다. 하지만 다수의 클라이언트가 요청을 하는 경우, 매번 `Connection`객체를 각각 생성해야 하기 때문에 생성에 대한 소요 시간이 많아진다. 
+
+### Connection Pool
+
+![3](https://user-images.githubusercontent.com/43839938/97099041-54af2700-16c7-11eb-8b01-b4971dafc7bf.png)
+
+`Connection Pool`은 `단순 Multiple Connection` 의 문제를 해결할 수 있다. `Connection Pool`은 미리 일정량의 `Connection` 객체를 생성하고 `Pool`이라는 공간에 저장한다. DB에 `Connection`해야하는 클라이언트 요청이 들어올 때, 해당 `Pool`에 가서 이미 생성되어 있고, 사용가능한 `Connection`을 찾아 DB에 접근한다. 
+
+npm mysql2 문서에는 다음과 같이 설명되어 있다. 
+
+```
+Connection pools help **reduce the time** **spent connecting to the MySQL server by reusing a previous connection**, leaving them open instead of closing when you are done with them.
+
+This improves the latency of queries as you avoid all of the overhead that comes with establishing a new connection.
+```
```

Comment:
```suggestion
> Connection pools help **reduce the time** **spent connecting to the MySQL server by reusing a previous connection**, leaving them open instead of closing when you are done with them.
>
> This improves the latency of queries as you avoid all of the overhead that comes with establishing a new connection.
```
인용문을 사용하는게 더 좋아 보일 것 같습니다.

Reply:
반영합니다 👍🏼